### DISCUSSION TITLE
custom metrics

### DISCUSSION THREAD

**Discussion by mirisu2** [labstack/echo-contrib#108] [Q&A]
Hi,
is it possible to add custom gauge metric for some function/endpoint? If so, how to make it? For example I would like to run goroutine which would make some action and set value for that metric.
Unfortunately, the documentation doesn't have any examples.

**Comment by aldas** [comment]
Here https://github.com/labstack/echo-contrib/blob/fdc9c9a0395a6abc42df8d9aadc1071f0a40ad2c/echoprometheus/prometheus_test.go#L222

is example that creates new prometheus counter and adds it to registry. In your use-case you do not need that `AfterNext` part.

```go
func TestMiddlewareConfig_AfterNextFuncs(t *testing.T) {
	e := echo.New()

	customRegistry := prometheus.NewRegistry()
	customCounter := prometheus.NewCounter(
		prometheus.CounterOpts{
			Name: "custom_requests_total",
			Help: "How many HTTP requests processed, partitioned by status code and HTTP method.",
		},
	)
	if err := customRegistry.Register(customCounter); err != nil {
		t.Fatal(err)
	}

	e.Use(NewMiddlewareWithConfig(MiddlewareConfig{
		AfterNext: func(c echo.Context, err error) {
			customCounter.Inc() // use our custom metric in middleware
		},
		Registerer: customRegistry,
	}))
	e.GET("/metrics", NewHandlerWithConfig(HandlerConfig{Gatherer: customRegistry}))

	e.GET("/ok", func(c echo.Context) error {
		return c.JSON(http.StatusOK, "OK")
	})

	assert.Equal(t, http.StatusOK, request(e, "/ok"))

	body, code := requestBody(e, "/metrics")
	assert.Equal(t, http.StatusOK, code)
	assert.Contains(t, body, `custom_requests_total 1`)
}
```

**Reply by mirisu2** [reply]
@aldas Thank you for reply. I also found [this](https://github.com/labstack/echo/discussions/1892#discussioncomment-855570).

**Reply by aldas** [reply]
Note for future readers:

```go
	myCounter := prometheus.NewGauge(prometheus.GaugeOpts{
		Name:        "my_handler_executions",
		Help:        "Counts executions of my handler function.",
		ConstLabels: prometheus.Labels{"version": "1234"},
	})
	if err := prometheus.Register(myCounter); err != nil {
		log.Fatal(err)
	}
```

creates new prometheus metric and registers it with default (global) metrics registry. This approach works also with newer echo prometheus middleware if you do not provide custom registry.
